doc/vlan_stacking/vlan_stacking_HLD.md

```diff
@@ -0,0 +1,635 @@
+# Overview
+
+This document describes the design details of the VLAN stacking feature. VLAN stacking is a feature designed for service providers who carry traffic of multiple customers across  their networks and are required to maintain the VLAN and Layer 2 protocol configurations of each customer without impacting the traffic of other customers. VLAN stacking include QinQ feature and VLAN Translation feature.
+
+IEEE 802.1Q tunneling (QinQ tunneling) uses a single Service Provider VLAN (SPVLAN) for customers who have multiple VLANs. Customer VLAN IDs are preserved and traffic from different customers is segregated within the service provider’s network even when they use the same customer-specific VLAN IDs. QinQ tunneling expands VLAN space by using a VLAN-in-VLAN hierarchy, preserving the customer’s original tagged packets, and adding SPVLAN tags to each frame (also called double tagging).
+
+QinQ tunneling uses double tagging to preserve the customer’s VLAN tags on traffic crossing the service provider’s network. However, if any switch in the path crossing the service provider’s network does not support this feature, then the switches directly connected to that device can be configured to swap the customer’s VLAN ID with the service provider’s VLAN ID for upstream traffic, or the service provider’s VLAN ID with the customer’s VLAN ID for downstream traffic.
+
+## VLAN Stacking Deployment Use Cases
+
+### QinQ Use Case
+
+The following example maps C-VLAN 10 to S-VLAN 100, C-VLAN 20 to S-VLAN 200 and C-VLAN 30 to S-VLAN 300 for ingress traffic on port 1 of Switches A and B.
+
+Figure 1. Mapping QinQ Service VLAN to Customer VLAN
+
+![QinQ use case](./images/QinQ_usecase.png)
+
+### VLAN Translation Use Case
+
+For example, assume that the upstream switch does not support QinQ tunneling. Select Port 1, and set the Old VLAN to 10 and the New VLAN to 100 to map VLAN 10 to VLAN 100 for upstream traffic entering port 1, and VLAN 100 to VLAN 10 for downstream traffic leaving port 1 as shown below.
+
+Figure 2. Configuring VLAN Translation
+
+![VLNA translate use case](./images/vlan_xlate_usecase.png)
+
+# Architecture Design
+
+The overall SONiC architecture will not be changed and no new sub-modules will be introduced.
+
+# Requirements
+
+* User **_shall_** be able to configure VLAN stacking on each Ethernet port and and channel port interface.
+* A VLAN stacking configuration describes how to determine the service that the specific traffic belongs to and what rewrite action **_will_** be done for the frames. Each configuration include the following:
+  * Service - A VLAN ID that the specified traffic **_will_** be applied to. Range 1 ~ 4094.
+  * Direction - Traffic direction.
+  * Action - The rewrite action for the frames, support the following actions:
+    * push
+      * Apply for all tagged frames, single or double tagged frames, at ingress direction.
+      * Add an outer VLAN with TPID=8100, COS=0, VLAN=<specified-S-VLAN> (range 1 ~ 4094).
```

Comment:
Any reason TPID=8100 was picked in place of IEEE defined 88a8. We should add details about CoS behavior. Is this always a fixed value or plan to support copy from inner etc..

Reply:
1. Currently we use the TPID=8100 as many other product do. This value can changed by predefine value.
2. Will write the cos behavior in this HLD.

Reply:
I assume we'll have an option to configure TPID value as 0x8100 (802.1Q) or 0x88a8 (802.1ad)

Reply:
Currently, it do not have option for this, it's also can add this to the HLD.

Reply:
It will add to the future work, below will add to the commit.
# Future Work

The values of TPID and COS in the adding outer VLAN are configurable. For example, TPID may be configured as 0x88a8 which is defined in IEEE or other values. For the COS value, it may be able to configure to extract from the COS value of the original VLAN.

Reply:
Ok, thanks!